ui/raidboss/emulator/data/AnalyzedEncounter.ts

```diff
@@ -0,0 +1,220 @@
+    }
+
+    const timelineUI = new RaidEmulatorAnalysisTimelineUI(this.options);
+    const timelineController =
+        new RaidEmulatorTimelineController(this.options, timelineUI, raidbossFileData);
+    timelineController.bindTo(this.emulator);
+    const popupText = new PopupTextAnalysis(
+        this.options, new TimelineLoader(timelineController), raidbossFileData);
+
+    const generator = new PopupTextGenerator(popupText);
+    timelineUI.SetPopupTextInterface(generator);
+
+    timelineController.SetPopupTextInterface(generator);
+
+    this.selectPerspective(id);
+
+    if (timelineController.activeTimeline) {
+      timelineController.activeTimeline.SetTrigger((trigger: LooseTrigger, matches) => {
+        const currentLine = this.encounter.logLines[currentLogIndex];
+
+        if (!currentLine)
+          throw new UnreachableCode();
+
+        popupText.OnTrigger(trigger, matches, currentLine.timestamp);
+
+        const triggerHelper =
+          popupText._onTriggerInternalGetHelper(trigger, {}, currentLine.timestamp);
+
+        if (!triggerHelper.resolver)
+          throw new UnreachableCode();
+
+        triggerHelper.resolver.status.finalData = EmulatorCommon.cloneData(popupText.getData());
+
+        if (!popupText.callback)
+          throw new UnreachableCode();
+
+        popupText.callback(
+            currentLine,
+            triggerHelper,
+            triggerHelper.resolver.status,
+            triggerHelper.resolver.status.finalData);
+      });
+    }
+
+    popupText.callback = (log, triggerHelper, currentTriggerStatus) => {
+      const perspective = this.perspectives[id];
+      if (!perspective || !triggerHelper)
+        throw new UnreachableCode();
+
+      const delay = currentTriggerStatus.delay ?? 0;
+
+      perspective.triggers.push({
+        triggerHelper: triggerHelper,
+        status: currentTriggerStatus,
+        logLine: log,
+        resolvedOffset: (log.timestamp - this.encounter.startTimestamp) +
+          (delay * 1000),
+      });
+    };
+    popupText.triggerResolvers = [];
+
+    this.perspectives[id] = {
+      initialData: EmulatorCommon.cloneData(popupText.getData(), []),
+      triggers: [],
+      finalData: popupText.getData(),
+    };
+
+    for (; currentLogIndex < this.encounter.logLines.length; ++currentLogIndex) {
+      const log = this.encounter.logLines[currentLogIndex];
+      if (!log)
+        throw new UnreachableCode();
+      await this.dispatch('analyzeLine', log);
+
+      const combatant = this.encounter?.combatantTracker?.combatants[id];
+
+      if (combatant && combatant.hasState(log.timestamp))
+        this.updateState(combatant, log.timestamp);
+
+      const event = {
+        type: 'onLogEvent',
+        detail: {
+          logs: [log],
+        },
+      };
+
+      await popupText.OnLog(event as EmulatorLogEvent);
+      await popupText.OnNetLog(event as EmulatorNetworkLogEvent);
+      timelineController.OnLogEvent(event as EmulatorLogEvent);
```

Comment:
I would like to avoid these as well.  I tried looking at the definitions of LogEvent and EmulatorLogEvent and I'm a little bit confused how to disentangle it.  Redefining logs as two totally different types is a little bit suspicious.  I'll take a look tomorrow when I have a little bit more brain.

Reply:
I could do some refactoring to avoid this but I'd have to modify all of the following files to compensate:

- raidemulator.js
- AnalyzedEncounter.ts
- PopupTextAnalysis.ts
- RaidEmulatorPopupText.js
- RaidEmulatorTimelineController.ts
- StubbedPopupText.ts

Basically I'd stub out `OnLog` and `OnNetLog` and equivalents in the parent classes, and replace existing emulator overrides for `OnLog` and `OnNetLog` with new `OnEmulatorLog` calls. Do you want me to do that as part of this PR?

Reply:
I've done this refactor with the latest commit. Please re-review and let me know if this is a better solution.